icechunk/src/session.rs

```diff
@@ -1445,15 +1448,47 @@ impl Session {
                 snap_id.clone(),
             );
 
+            let finder = PathFinder::new(session.list_nodes(&Path::root()).await?);
             let mut fresh = self.change_set().fresh();
             std::mem::swap(self.change_set_mut()?, &mut fresh);
             let change_set = fresh;
             // TODO: this should probably execute in a worker thread
+
             match solver.solve(&tx_log, &session, change_set, self).await? {
-                ConflictResolution::Patched(patched_changeset) => {
+                ConflictResolution::Patched(mut patched_changeset) => {
                     trace!("Snapshot rebased");
-                    self.change_set = patched_changeset;
+
+                    // important to set this here, so that the `get_array` below gets the
+                    // *updated* node with updated shape and dimension_names
                     self.snapshot_id = snap_id;
+
+                    let mut new_splits: HashMap<NodeId, ManifestSplits> = HashMap::new();
+                    // we have rebased on top of `snap_id`, now we update splits to match the config in that snap.
+                    // `splits` would have changed if an array was resized. `flush` uses `splits` so if left unupdated,
+                    // we run the risk of losing data.
+                    for node_id in patched_changeset.arrays_with_chunk_changes() {
+                        let node_path = finder.find(node_id)?;
+                        // by now the conflict solver has solved any needed conflicts,
+                        // we grab the *updated* node to get the *updated* metadata (shape, dimension_names)
+                        if let NodeSnapshot {
+                            node_data: NodeData::Array { shape, dimension_names, .. },
+                            ..
+                        } = self.get_array(&node_path).await?
+                        {
+                            let new_size = session
+                                .config()
+                                .manifest()
+                                .splitting()
+                                .get_split_sizes(&node_path, &shape, &dimension_names);
+                            new_splits.insert(node_id.clone(), new_size);
+                        };
+                    }
+                    // now we update the changeset to be consistent with these new splits
+                    patched_changeset.replay_set_chunk_refs(&new_splits)?;
```

Comment:
This is expensive, can we do it only for the arrays that had a change to their splits?